### PR TITLE
Test registration and authentication workflow

### PR DESCRIPTION
**Pull request by botandrose**
Closes #38

### CODE REVIEW

**Comment by botandrose** [comment]
@mradamh @ageoldpun I've never designed an API before, so this is uncertain ground for me. What do you guys think about this registration and auth design?

**Comment by botandrose** [comment]
Tested this with my secret implementation that Paul is not allowed to see ;) so I'm more or less happy with it. Any feedback before I merge?

File: test/api/auth_test.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            password: "badpassword",
+          },
+        }]
+      })
+      response.json_body.must_equal_json({
+        errors: [{
+          title: "Invalid authentication credentials",
+        }]
+      })
+      response.code.must_equal 403
```
Comment (mradamh) [inline test/api/auth_test.rb]: IMO this should be a 401.  403 means you have successfully authenticated (i.e. the server know who you are) but you are forbidden from the requested resource.

File: test/api/auth_test.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+          type: "auth",
+          attributes: {
+            email: "micah@botandrose.com",
+          },
+        }]
+      })
+      response.code.must_equal 200
+
+      response = get("/api/greet", "Authorization" => "Bearer xyz.789")
+      response.code.must_equal 403
```
Comment (mradamh) [inline test/api/auth_test.rb]: same here.  401.

File: test/api/auth_test.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      token.must_match /\A\w+\.\w+\.\w+\Z/
+      json_body.must_equal_json({
+        data: [{
+          type: "auth",
+          attributes: {
+            email: "micah@botandrose.com",
+          },
+        }]
+      })
+      response.code.must_equal 200
```
Comment (mradamh) [inline test/api/auth_test.rb]: I don't know ruby at all.  i assume L73 means there is a `token` attribute on the response that is independently verified and then deleted from `json_body`

Just IMO, but the mutation makes it harder to read.

File: test/support/schema.sql [hunk 4]
```diff
@@ -14,3 +14,15 @@ CREATE TABLE `clients` (
...
   PRIMARY KEY (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
+
+DROP TABLE IF EXISTS `users`;
+CREATE TABLE `users` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `name` varchar(255) DEFAULT NULL,
+  `email` varchar(255) DEFAULT NULL,
+  `crypted_password` varchar(255) DEFAULT NULL,
+  `password_salt` varchar(255) DEFAULT NULL,
```
Comment (mradamh) [inline test/support/schema.sql]: I've only ever used bcrypt for auth with stores the salt encoded with the password hash, not that it necessarily matters.

File: test/api/auth_test.rb (hunk 1 above)
Comment (botandrose) [inline test/api/auth_test.rb]: Ah, yes, good distinction. Classic authentication and authorization conflation.

File: test/api/auth_test.rb (hunk 3 above)
Comment (botandrose) [inline test/api/auth_test.rb]: You are correct. I hate this, and also the similar ugly bit above for validating timestamps. I can think of two options for removing this:

1. Have a default secret key that can be overridden by an ENV variable, and assert on the exact encoded JWT string. Pretty damn opaque, but still more readable than this garbage.
2. Extend must_equal_json to accept arbitrary regexen as attributes, something like:

```ruby
response.json_body.must_equal_json({
  data: [{
    type: "auth",
    attributes: {
      email: "micah@botandrose.com",
      token: /\A\w+\.\w+\.\w+\Z/,
    },
  }]
})
```

I think I'm leaning towards the latter, even though its more work. What do you think?

File: test/api/auth_test.rb (hunk 2 above)
Comment (botandrose) [inline test/api/auth_test.rb]: :+1:

File: test/support/schema.sql (hunk 4 above)
Comment (botandrose) [inline test/support/schema.sql]: Yeah, I wrote eboshi back in 2008, so the schema reflects that somewhat. I think this was before bcrypt hit the scene, and sha256 was still all the rage.

**Comment by botandrose** [comment]
Thanks for the review, @mradamh. Fixed the status codes, and gonna merge now. Splitting off the ugly destructuring+regex garbage into a new issue.

**Pull request merged**